core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1617,10 +1617,20 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     // may actually be ahead of the current producer state end offset (which corresponds to the log end offset),
     // we manually override the state offset here prior to taking the snapshot.
     producerStateManager.updateMapEndOffset(newSegment.baseOffset)
-    producerStateManager.takeSnapshot()
+    // We avoid potentially-costly fsync call, since we acquire UnifiedLog#lock here
+    // which could block subsequent produces in the meantime.
+    // flush is done in the scheduler thread along with segment flushing below
+    val maybeSnapshot = producerStateManager.takeSnapshot(false)
     updateHighWatermarkWithLogEndOffset()
     // Schedule an asynchronous flush of the old segment
-    scheduler.scheduleOnce("flush-log", () => flushUptoOffsetExclusive(newSegment.baseOffset))
+    scheduler.scheduleOnce("flush-log", () => {
+      maybeSnapshot.ifPresent(f => {
+        maybeHandleIOException(s"Error while deleting producer state snapshot $f for $topicPartition in dir ${dir.getParent}") {
+          Utils.flushFileIfExists(f.toPath)
+        }
+      })
+      flushUptoOffsetExclusive(newSegment.baseOffset)
```

Comment:
Is it possible to add a test to verify that the recovery point is only advanced after the producer state has been flushed to disk?